Update CAP-0033 (#751)
stellar/stellar-protocol: core/cap-0033.md

  create highly usable products.

## Abstract
We introduce the sponsoring-future-reserves relation, in which an account (the
sponsoring account) pays any reserve that another account (the sponsored account)
would have to pay. This relation is initiated by `BeginSponsoringFutureReservesOp`,
where the sponsoring account is the source account, and is terminated by

> `EndSponsoringFutureReservesOp`, where the sponsored account is the source

Reviewer: I was just reading this a minute ago while writing up more examples for the docs and thought it may be worth highlighting the difference between Begin/End here:

> This relation is initiated by `BeginSponsoringFutureReservesOp`, where the **sponsoring** account is the source account, and is terminated by `EndSponsoringFutureReservesOp`, where the **sponsored** account is the source account.

It's very easy to otherwise overlook that tiny _textual_ but large _conceptual_ difference.

> where the sponsoring account is the source account, and is terminated by
> `EndSponsoringFutureReservesOp`, where the sponsored account is the source

Reviewer 2: ```suggestion
where the **sponsoring** account is the source account, and is terminated by
`EndSponsoringFutureReservesOp`, where the **sponsored** account is the source
```

  is not sponsoring-future-reserves for any account. Specifically, this
  prevents an account from sponsoring-future-reserves for itself.
- If an account `A` is sponsoring-future-reserves for an account `B`, then `A`
  exists. The converse need not be true (this facilitates creating a sponsored
  account).

#### Sponsorship Invariants
There are three invariants related to sponsorships:
- (Global) The sum of all `numSponsoring` is equal to the sum of all
  `numSponsored` plus the sum of all `ClaimableBalance.claimants.size()`.
- (Local) For an account `A`, `numSponsoring` is equal to the count of all
  `LedgerEntry le` with `le.ext.v1().sponsoringID == A` weighted by the number
  of reserves required for that `LedgerEntry` (2 for `ACCOUNT`, `claimants.size()`
  for `CLAIMABLE_BALANCE`, 1 otherwise) plus the count of all `A` in
  `signerSponsoringIDs`.
- (Local) For an account `A`, `numSponsored` is equal to the count of all
  `LedgerEntry le` which are sub-entries of `A` (including `A` itself) that have
  `le.ext.v1().sponsoringID != null` weighted by the number
  of reserves required for that `LedgerEntry` (2 for `ACCOUNT`, `claimants.size()`

>   for `CLAIMABLE_BALANCE`, 1 otherwise) plus the count of all `s` in

Reviewer 2: Claimable balances don't affect `numSponsored`. We should clarify that here.

>   `le.ext.v1().sponsoringID != null` weighted by the number
>   of reserves required for that `LedgerEntry` (2 for `ACCOUNT`, `claimants.size()`
>   for `CLAIMABLE_BALANCE`, 1 otherwise) plus the count of all `s` in
>   `A.ext.v1().ext.v2().signerSponsoringIDs` with `s != null`.

Reviewer 2: ```suggestion
  `le.ext.v1().sponsoringID != null` weighted by the number of reserves 
  required for that `LedgerEntry` (2 for `ACCOUNT`, 1 otherwise) plus the count of all `s` 
  in `A.ext.v1().ext.v2().signerSponsoringIDs` with `s != null`. Note that
  `Claimable balances` are not sub-entries, so they do not affect `numSponsored`.
```

> #### Reserve Requirement
> No operation can take the native balance of an account below
> `(2 + numSubEntries + numSponsoring - numSponsored) * baseReserve + liabilities.selling`.

Reviewer 2: ```suggestion
```

`EndSponsoringFutureReservesOp` requires medium threshold.

#### RevokeSponsorshipOp
`RevokeSponsorshipOp` allows a ledger entry or signer

- that is not sponsored to be sponsored,
- that is sponsored to change sponsors, or
- that is sponsored to no longer be sponsored.

To check validity of `RevokeSponsorshipOp op`:
```
If ledgerVersion != 14
    If op.type() == REVOKE_SPONSORSHIP_LEDGER_ENTRY
        If op.ledgerKey().type() == TRUSTLINE
            Let tl = op.ledgerKey().trustLine()
            If tl.asset is not valid
                Fail with REVOKE_SPONSORSHIP_DOES_NOT_EXIST
            If tl.asset.type() == ASSET_TYPE_NATIVE
                Fail with REVOKE_SPONSORSHIP_DOES_NOT_EXIST
            If tl.accountID == getIssuer(tl.asset)
                Fail with REVOKE_SPONSORSHIP_DOES_NOT_EXIST
        If op.ledgerKey().type() == OFFER
            If op.ledgerKey().offer().offerID <= 0
                Fail with REVOKE_SPONSORSHIP_DOES_NOT_EXIST
        If op.ledgerKey().type() == DATA
            Let de = op.ledgerKey().data()

>             If data.dataName is empty
>                 Fail with REVOKE_SPONSORSHIP_DOES_NOT_EXIST
>             If data.dataName is not valid

Reviewer 2: ```suggestion
            If de.dataName is empty
                Fail with REVOKE_SPONSORSHIP_DOES_NOT_EXIST
            If de.dataName is not valid
```

    getNumSponsoring(newSponsor) += 1

    Let i be the index of op.signer().signerKey in acc.signers
    Let sid = acc.ext.v1().ext.v2().signerSponsoringIDs[i]
    Load sid as oldSponsor
    getNumSponsoring(oldSponsor) -= 1

    acc.ext.v1().ext.v2().signerSponsoringIDs[i] = newSponsor.accountID
Else if wasSponsored and !willBeSponsored
    // Stop sponsoring something that was sponsored
    If getAvailableBalance(acc) < baseReserve
        Fail with REVOKE_SPONSORSHIP_LOW_RESERVE
    getNumSponsored(acc) -= 1

    Let i be the index of op.signer().signerKey in acc.signers
    Let sid = acc.ext.v1().ext.v2().signerSponsoringIDs[i]
    Load sid as oldSponsor
    getNumSponsoring(oldSponsor) -= 1

    Clear acc.ext.v1().ext.v2().signerSponsoringIDs[i]
Else if !wasSponsored and willBeSponsored
    // Sponsor something that wasn't sponsored
    Load account sponsoring future reserves for op.sourceAccount as newSponsor
    If getAvailableBalance(newSponsor) < baseReserve
        Fail with REVOKE_SPONSORSHIP_LOW_RESERVE
    If getNumSponsoring(newSponsor) > UINT32_MAX - 1
        Fail with opTOO_MANY_SPONSORING
    getNumSponsoring(newSponsor) += 1

>     getNumSponsored(acc) += mult

Reviewer 2: ```suggestion
    getNumSponsored(acc) += 1
```
